### PR TITLE
Report stencil argument intents, update fv3core docstrings

### PR DESCRIPTION
**Pull request by mcgibbon**
## Purpose

This PR adds a script to generate a report with all stencil call signatures showing in/out/inout intents, and makes use of its output to update fv3core (outside of remapping, which will be done in another PR).

## Code changes:

- added an example script in `driver` which reports the call signature of each stencil, and aggregated call signatures for classes containing stencils
- added and updated all docstrings outside of remapping in fv3core with in/out/inout based on report and manual inspection
- re-ordered object creation and stencil function definitions to match call order
- inlined some helper methods in fv3core
- added `Driver` and `DriverConfig` to `pace.driver` namespace

## Infrastructure changes:

- enabled mypy for driver in pre-commit config

## Checklist
Before submitting this PR, please make sure:

- [x] You have followed the coding standards guidelines established at [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--BD7zigBMAhMZAPkeNENeuU2UAg-IlsYffZgTwyKEylty7NhY).
- [x] Docstrings and type hints are added to new and updated routines, as appropriate
- [ ] All relevant documentation has been updated or added (e.g. README, CONTRIBUTING docs)
- [x] For each public change and fix in `pace-util`, HISTORY has been updated
- [x] Unit tests are added or updated for non-stencil code changes

### CODE REVIEW

File: fv3core/fv3core/stencils/updatedzd.py [hunk 1]
```diff
@@ -347,15 +348,15 @@ def __call__(
...
-            surface_height: height of surface (in)
-            height: height defined on layer interfaces (inout)
-            courant_number_x: Courant number in x-direction defined on cell centers (in)
-            courant_number_y: Courant number in y-direction defined on cell centers (in)
-            x_area_flux: Area flux in x-direction defined on cell centers (in)
-            y_area_flux: Area flux in y-direction defined on cell centers (in)
-            ws: lowest layer vertical velocity implied by horizontal motion
-                over topography, in units of [height units] / second (out)
-            dt: timestep over which input fluxes have been computed, in seconds
+            surface_height (in): height of surface (in)
```
Comment (oelbert) [inline fv3core/fv3core/stencils/updatedzd.py:351]:
```suggestion
            surface_height (in): height of surface
```

File: fv3core/fv3core/stencils/updatedzd.py [hunk 2]
```diff
@@ -347,15 +348,15 @@ def __call__(
...
+            height (inout): height defined on layer interfaces (inout)
```
Comment (oelbert) [inline fv3core/fv3core/stencils/updatedzd.py:352]:
```suggestion
            height (inout): height defined on layer interfaces
```

File: fv3core/fv3core/stencils/temperature_adjust.py [hunk 3]
```diff
@@ -19,13 +19,13 @@ def compute_pkz_tempadjust(
...
-        heat_source: heat source from vorticity damping implied by
-            energy conservation (in)
-        pt: Air potential temperature (inout)
-        pkz: Layer mean pressure raised to the power of Kappa (in)
+        delp (in): Pressure thickness of atmosphere layers
+        delz (in): Vertical thickness of atmosphere layers
+        cappa (in): Power to raise pressure to
+        heat_source (in): heat source from vorticity damping implied by
+            energy conservation
+        pt (inout): Air potential temperature (inout)
```
Comment (oelbert) [inline fv3core/fv3core/stencils/temperature_adjust.py:27]:
```suggestion
        pt (inout): Air potential temperature
```

File: fv3core/fv3core/stencils/temperature_adjust.py [hunk 4]
```diff
@@ -19,13 +19,13 @@ def compute_pkz_tempadjust(
...
+        pkz (out): Layer mean pressure raised to the power of Kappa
         delta_time_factor: scaled time step (in)
```
Comment (oelbert) [inline fv3core/fv3core/stencils/temperature_adjust.py:29]:
```suggestion
        delta_time_factor (in): scaled time step
```

File: fv3core/fv3core/stencils/sim1_solver.py [hunk 5]
```diff
@@ -140,20 +152,20 @@ def __call__(
...
+        vertical velocity and pressure perturbations.
 
+        Args:
+          dt (in): timstep in seconds of solver
+          gm (in): ?? 1 / (1 - cappa)
+          cp3 (in): cappa
+          pe (out): full hydrostatic pressure
+          dm (in): delta in pressure / g
+          pm (in): ?? ratio of change in layer pressure without condensates
+          pem (in): recomputed pressure using ptop and delp(
```
Comment (oelbert) [inline fv3core/fv3core/stencils/sim1_solver.py:164]:
```suggestion
          pem (in): recomputed pressure using ptop and delp
```

File: fv3core/fv3core/stencils/a2b_ord4.py [hunk 6]
```diff
@@ -608,14 +682,24 @@ def _exclude_tile_edges(self, origin, domain, dims=("x", "y")):
...
+        """
+        Converts qin from A-grid to B-grid in qout.
+
+        If initialized with replace=True, qin is also updated to the B grid.
 
         Args:
-            qin: Input on A-grid (inout)
-            qout: Output on B-grid (inout)
+            qin (inout): Input on A-grid (intent=in if replace=false)
+            qout (out): Output on B-grid (out)
```
Comment (oelbert) [inline fv3core/fv3core/stencils/a2b_ord4.py:692]:
```suggestion
            qout (out): Output on B-grid
```

File: fv3core/fv3core/stencils/c_sw.py [hunk 7]
```diff
@@ -588,14 +673,15 @@ def __call__(
...
-            vt: v * dy (inout)
-            divgd: D-grid horizontal divergence (inout)
-            omga: Vertical pressure velocity (inout)
+            ua: A-grid x-velocity (inout)
+            va: A-grid y-velocity (inout)
+            ut: u * dx (out)
+            vt: v * dy (out)
+            divgd: D-grid horizontal divergence (out)
+            omga: Vertical pressure velocity (out)
             dt2: Half a model timestep in seconds (in)
```
Comment (oelbert) [inline fv3core/fv3core/stencils/c_sw.py:682]: The layouts of these are inconsistent with the other in/out documentation

File: fv3core/fv3core/stencils/d_sw.py [hunk 8]
```diff
@@ -886,45 +990,54 @@ def __call__(
...
-            cry: local courant number in the y direction (inout)
-            xfx: flux of area in x-direction, in units of m^2 (in)
-            yfx: flux of area in y-direction, in units of m^2 (in)
-            q_con: total condensate mixing ratio (inout)
-            zh: geopotential height defined on layer interfaces (in)
-            heat_source:  accumulated heat source (inout)
-            diss_est: dissipation estimate (inout)
-            dt: acoustic timestep in seconds (in)
+            mfx (inout): accumulated x mass flux (inout)
+            mfy (inout): accumulated y mass flux (inout)
```
Comment (oelbert) [inline fv3core/fv3core/stencils/d_sw.py:1014]: More duplication

**Review by oelbert — CHANGES_REQUESTED** [review]
A few small changes, mostly just cleanup for duplicated in/out hinting. I didn't double-check the signatures, but I think it's ok and I could go back over it if you want.

**Comment by mcgibbon** [comment]
launch jenkins

File: fv3core/fv3core/stencils/divergence_damping.py [hunk 9]
```diff
@@ -413,76 +498,14 @@ def __call__(
...
-            v_contra_dxc,
-            ke,
-            d2_bg,
-            self._da_min_c,
-            self._dddmp,
-            dt,
-        )
-
     def _vorticity_calc(self, wk, vort, delpc, dt):
+        # this function is kept because it has a translate test, if its
```
[the remark this answers is not in the document]
Comment (oelbert) [reply fv3core/fv3core/stencils/divergence_damping.py:502]: Yeah, I think that would be better

File: fv3core/fv3core/stencils/c_sw.py (hunk 7 above)
Comment (mcgibbon) [reply fv3core/fv3core/stencils/c_sw.py]: Done.

**Review by oelbert — APPROVED** [review]
Looks good!

**Pull request merged**